hub/config/src/main/java/uk/gov/ida/hub/config/data/S3ConfigSource.java

```diff
@@ -49,33 +58,61 @@ public static S3ConfigSource setupS3ConfigSource(ConfigConfiguration configConfi
-                new GetObjectRequest(configConfiguration.getSelfService().getS3BucketName(),
-                configConfiguration.getSelfService().getS3ObjectKey()));
-        InputStream s3ObjectStream =  fullObject.getObjectContent();
+    public RemoteConfigCollection getRemoteConfig() throws IOException {
+        try {
+            // Object has never been retrieved get it
+            if (lastModifiedDate == null) {
+                lastAccessedTime = System.currentTimeMillis();
+                getObjectData();
+            }
+            // If the cache has timed out see if the object has changed
+            else if (lastAccessedTime + cacheLengthInMillis < System.currentTimeMillis()) {
+                lastAccessedTime = System.currentTimeMillis();
+                // Get Object meta data... see if it has been changed.
+                if (this.lastModifiedDate.before(s3Object.getObjectMetadata().getLastModified())) {
+                    // If the object has changed get the object from s3
+                    getObjectData();
+                }
+            }
+        } catch (IOException e) {
+            if(this.currentConfig != null)
+                return this.currentConfig;
+            else
+                LOG.error("A error has occurred... There is no cached remote config available, falling back to federated config");
+                throw e;
+        }
+
+        return this.currentConfig;
+    }
+
+    private void getObjectData() throws IOException {
+        this.lastModifiedDate = s3Object.getObjectMetadata().getLastModified();
+        S3ObjectInputStream s3ObjectStream = s3Object.getObjectContent();
         ObjectMapper om = new ObjectMapper();
         DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss");
         om.setDateFormat(df);
         SimpleModule module = new SimpleModule();
         module.addDeserializer(RemoteConfigCollection.class, new RemoteConfigCollectionDeserializer());
         om.registerModule(module);
         try {
-            return om.readValue(s3ObjectStream, RemoteConfigCollection.class);
+            this.currentConfig = om.readValue(s3ObjectStream, RemoteConfigCollection.class);
         } catch(IOException e) {
-            LOG.error("An error occured trying to get or process object {} from S3 Bucket {}",
+            LOG.warn("An error occurred trying to get or process object {} from S3 Bucket {}",
                     configConfiguration.getSelfService().getS3ObjectKey(),
                     configConfiguration.getSelfService().getS3BucketName(), e);
-            throw new RuntimeException(e);
+            throw e;
         } finally {
-            if(fullObject != null) {
+            if(this.s3Object != null) {
                 try {
-                    fullObject.close();
+                    this.s3Object.close();
                 } catch(IOException e) {
                     throw new RuntimeException(e);
                 }
             }
```

Comment:
I think we could do a better job of the exception handling here. If we're going to throw the IOException anyway, they we shouldn't need all this.